### PR TITLE
Three code-level optimization to Iterator::Next()

### PR DESCRIPTION
**Pull request by siying**
Summary:
Three small optimizations:
(1) iter_->IsKeyPinned() shouldn't be called if read_options.pin_data is not true. This may trigger function call all the way down the iterator tree.
(2) reuse the iterator key object in DBIter::FindNextUserEntryInternal(). The constructor of the class has some overheads.
(3) Move the switching direction logic in MergingIterator::Next() to a separate function.

These three in total improves readseq performance by about 3% in my benchmark setting.

Test Plan: Run all existing tests.

### CODE REVIEW

File: db/db_iter.cc [hunk 1]
```diff
@@ -264,6 +264,7 @@ class DBIter final: public Iterator {
 
   Status status_;
   IterKey saved_key_;
+  ParsedInternalKey tmp_ikey_;
```
Comment (sagar0) [inline db/db_iter.cc]: I suggest to rename this to something more meaningful, as this is a class field and not a function-local variable.

Comment (siying) [reply db/db_iter.cc]: Do you have any suggestion? Or I can add comments explaining it.

Comment (sagar0) [reply db/db_iter.cc]: May be call it just `ikey_`, and add a comment about the overhead.

**Comment by siying** [comment]
Addressed the comment.

**Pull request closed without merging**